backend/jottacloud/jottacloud.go

```diff
@@ -0,0 +1,799 @@
+// LastModified returned in the http headers
+func (o *Object) ModTime() time.Time {
+	err := o.readMetaData()
+	if err != nil {
+		fs.Logf(o, "Failed to read metadata: %v", err)
+		return time.Now()
+	}
+	return o.modTime
+}
+
+// SetModTime sets the modification time of the local fs object
+func (o *Object) SetModTime(modTime time.Time) error {
+	return fs.ErrorCantSetModTime
+}
+
+// Storable returns a boolean showing whether this object storable
+func (o *Object) Storable() bool {
+	return true
+}
+
+// Open an object for read
+func (o *Object) Open(options ...fs.OpenOption) (in io.ReadCloser, err error) {
+	fs.FixRangeOption(options, o.size)
+	var resp *http.Response
+	opts := rest.Opts{
+		Method:     "GET",
+		Path:       o.filePath(),
+		Parameters: url.Values{},
+		Options:    options,
+	}
+
+	opts.Parameters.Set("mode", "bin")
+
+	err = o.fs.pacer.Call(func() (bool, error) {
+		resp, err = o.fs.srv.Call(&opts)
+		return shouldRetry(resp, err)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, err
+}
+
+// Update the object with the contents of the io.Reader, modTime and size
+//
+// If existing is set then it updates the object rather than creating a new one
+//
+// The new object may have been created if an error is returned
+func (o *Object) Update(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (err error) {
+	size := src.Size()
+
+	var resp *http.Response
+	var result api.JottaFile
+	opts := rest.Opts{
+		Method:        "POST",
+		Path:          o.filePath(),
+		Body:          in,
+		ContentType:   "application/octet-stream",
+		ContentLength: &size,
+		ExtraHeaders:  make(map[string]string),
+		Parameters:    url.Values{},
+	}
+
+	md5, err := src.Hash(hash.MD5)
+	if err != nil {
+		opts.ExtraHeaders["JMd5"] = md5
+		opts.Parameters.Set("cphash", md5)
+
+		// TODO ? : If MD5 unsupported calculate MD5 ourself
+	}
+
+	opts.ExtraHeaders["JSize"] = strconv.FormatInt(size, 10)
+	//opts.ExtraHeaders["JCreated"] =
+	opts.ExtraHeaders["JModified"] = api.Time(src.ModTime()).String()
+
+	// Parameters observed in other implementations
+	//opts.ExtraHeaders["X-Jfs-DeviceName"] = "Jotta"
+	//opts.ExtraHeaders["X-Jfs-Devicename-Base64"] = ""
+	//opts.ExtraHeaders["X-Jftp-Version"] = "2.4" this appears to be the current version
+	//opts.ExtraHeaders["jx_csid"] = ""
+	//opts.ExtraHeaders["jx_lisence"] = ""
+
+	opts.Parameters.Set("umode", "nomultipart")
+
+	err = o.fs.pacer.CallNoRetry(func() (bool, error) {
+		resp, err = o.fs.srv.CallXML(&opts, nil, &result)
+		return shouldRetry(resp, err)
+	})
+	if err != nil {
+		return err
+	}
+
+	// TODO: Check returned Metadata? Timeout on big uploads?
+	return o.setMetaData(&result)
+}
+
+// Remove an object
+func (o *Object) Remove() error {
+	opts := rest.Opts{
+		Method:     "POST",
+		Path:       o.filePath(),
+		Parameters: url.Values{},
+	}
+
+	opts.Parameters.Set("dl", "true")
+
+	return o.fs.pacer.Call(func() (bool, error) {
+		resp, err := o.fs.srv.Call(&opts)
```

Comment:
If you are using Call you need to explicitly close the resp.Body otherwise you'll leak a file descriptor - see its docs.

You can do this by using by using CallXML(&opts, nil, nil) or by setting NoResponse: true in opts.

Reply:
Missed that in the docs. Makes sense.